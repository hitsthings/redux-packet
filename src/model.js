import bundler from './bundler';
import { combineReducers } from 'redux';

/*
const loadUsersForGroup = group => ({
    type: 'LOAD_USERS',
    group,
    ....
});

const reducer = keyBy(action => action.type, action => action.group)(finalReducer);
const selector = groupSelector => keyedBy(() =>'LOAD_USERS', groupSelector)(finalSelector);

// arg -> payload -> (serialize) -> reducerKey == selectorKey <- (serialize) <- inputSelector
[
    {
        name: 'group',
        getId: group => group.id
    }
]

const {
    request,
    success,
    failure,
    reducer,
    selector
} = entityDuck(
    'LOAD_USERS',
    function getId(entity) {
        return entity.id;
    },
    function payloadCreator(group) {
        return { group };
    },
    [
        payload => payload.group
    ]
)(
    function leafReducer(state, action) {
        switch(action.status) {
            case 'pending':
                return {
                    ...state,
                    loading: true,
                };
            case 'succeeded':
                return {
                    ...state,
                    loading: false
                }
        }
    },
    function leafSelector(state) {

    }
);
*/

const keyBy = (...keySelectors) => leafReducer => (state = {}, action) => {
    const keys = [];
    let oldLeafState = state;
    for (let i = 0; i < keySelectors.length; i++) {
        const key = keySelectors[i](action);
        if (key == null) { // if any key is invalid, return the previous state
            return state;
        }
        keys.push(key);
        oldLeafState = oldLeafState && oldLeafState[key];
    }

    const newLeafState = leafReducer(oldLeafState, action);
    // return the old state if it would be unchanged
    if (oldLeafState === newLeafState) {
        return state;
    }

    const newState = { ...state };
    const lastKey = keys.pop();
    const leafState = keys.reduce((leafState, key) => {
        return (leafState[key] = { ...leafState[key] });
    }, newState);
    leafState[lastKey] = newLeafState;
    return newState;
};
const keyedBy = (...keySelectors) => leafSelector => (state, props) => {
    const keys = [];
    let leafState = state;
    for (let i = 0; i < keySelectors.length; i++) {
        if (leafState === null) {
            return leafSelector(undefined, props);
        }
        const key = keySelectors[i](action);
        if (key == null) {
            return leafSelector(undefined, props);
        }
        leafState = leafState[key];
    }
    return leafSelector(leafState, props);
};

const entityDuck = (entityType, getId) => {
    //TODO - support multiple action types that all modify the same entities state, but different requests state.
    return (type, payloadCreator, reducerKeys, getEntities, getErrors, getPagingInfo) => {
        const request = (...args) => ({
            type,
            entityType,
            status: 'pending',
            payload: payloadCreator(...args)
        });
        const success = (requestAction, payload) => ({
            type,
            entityType,
            status: 'succeeded',
            context: requestAction,
            payload
        });
        const failure = (requestAction, payload) => ({
            type,
            entityType,
            status: 'failed',
            context: requestAction,
            payload
        });
        const entitiesReducer = function (state = {}, action) {
            if (action.type === type && action.status === 'succeeded') {
                const entities = getEntities(action);
                if (0 === entities.length) {
                    return state;
                }
                const newState = { ...state };
                entities.forEach(entity => {
                    newState[getId(entity)] = entity;
                });
                return newState;
            }
        };
        const requestsReducer = keyBy(...reducerKeys)(function(entityState = {}, action) {
            switch (action.status) {
                case 'pending':
                    return {
                        ...entityState,
                        loading: true,
                    };
                case 'failed':
                    return {
                        ...entityState,
                        loading: false,
                        errors: getErrors(action),
                    };
                case 'succeeded':
                    return {
                        loading: false,
                        errors: null,
                        pagingInfo: getPagingInfo(action),
                        ids: (entityState.ids || []).concat(getEntities(action).map(entity => getId(entity))),
                    };
            }
        });
        const combinedReducer = combineReducers({
            entities: entitiesReducer,
            requests: requestsReducer,
        });
        const reducer = (state = {}, action) => {
            if (action.type !== type) {
                return state;
            }
            return combinedReducer(state, action);
        };
        const selector = (...keySelectors) => {
            if (keySelectors.length !== reducerKeys.length) {
                throw new Error("The number of provided input selectors doesn't match the expected number of context inputs.");
            }
            return keyedBy(() => 'requests', () => type, ...keySelectors)(makeLeafSelector());
        }
        return {
            request,
            success,
            failure,
            reducer, // need one per ENTITY
            selector, // need one per action type
        };
    };
};