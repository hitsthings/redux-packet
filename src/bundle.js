import { bindActionCreators } from "redux";

/*
//usage of bundle
bundler({
    fromContext: {
        selector: (state, contextA, contextB) => `itemsInContext`, // or a function returning this to make one per component
        actions: (contextA, contextB) => `actionsForContext`,
    }
})
const users = bundle({
    fromGroup: {
        selector: () => createStructuredSelector({
            users: usersForGroupSelector,
            usersLoading: usersLoadingForGroupSelector,
            usersAllLoaded: usersAllLoadedForGroupSelector,
            usersLoadingErrors: usersLoadingErrorsForGroupSelector,
        }),
        actions: (dispatch, group) => ({
            loadUsers: () => dispatch(loadUsersForGroup(group)),
            addUserToGroup: user => dispatch(addUserToGroup(user, group)),
        }), // OR a raw object like actions: { actionA: actionCreatorA, actionB: actionCreatorB }
    },
    byPermission: { ... }
})
*/

const memoizeFirstMapStateToProps = (selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext) => {
    if (executeContextSelectors) {
        let inner = (state, props) => {
            inner = (state, props) => selector(state, ...executeContextSelectors(props));
            if (props !== firstProps) { // nothing salvageable
                return inner(state, props); 
            }
            if (state !== firstState) { 
                return selector(state, ...firstContext); // same props, can memoize contexts
            }
            return firstValue;// same props and state, can return first value
        };
        return (state, props) => inner(state, props);
    }
    let inner = state => {
        inner = selector;
        return state === firstState ?
            firstValue :
            inner(state);
    };
    return state => inner(state);
};

const checkBundleDescriptor = (bundleDescriptor, name, calledFnName, argExpectation) => {
    if (typeof bundleDescriptor !== 'object' || bundleDescriptor === null) {
        throw new TypeError(`${calledFnName} expects ${argExpectation}an object, but ${name} was not.`);
    }
    if (bundleDescriptor.selector && typeof bundleDescriptor.selector !== 'function') {
        throw new TypeError(`${calledFnName} expects selector to be a function, but ${name}.selector was not.`);
    }
    if (bundleDescriptor.actions && (typeof bundleDescriptor.actions !== 'function' && typeof bundleDescriptor.actions !== 'object')) {
        throw new TypeError(`${calledFnName} expects actions to be a function or object, but ${name}.actions was not.`);
    }
};

const checkBundleMapAndGetNames = bundleMap => {
    if (typeof bundleMap !== 'object' || bundleMap === null) {
        throw new TypeError('bundleAll(bundleMap) expects bundleMap to be an object.');
    }
    if (Array.isArray(bundleMap)) {
        throw new TypeError('bundleAll(bundleMap) expects bundleMap to be an object, not an Array.');
    }
    const bundleNames = Object.keys(bundleMap);
    if (bundleNames.length === 0) {
        throw new TypeError('bundleAll(bundleMap) expected to receive an object with properties, but found no properties.')
    }
    return bundleNames;
};

const selectorExecutor = selectors => {
    selectors = selectors.map(selector => typeof selector === 'function' ? selector : () => selector);
    return props => selectors.map(selector => selector(props));
};

const internalBundle = (bundleDescriptor, name) => {
    const Bundle = (...contextSelectors) => {
        const hasContext = contextSelectors.length !== 0;
        const executeContextSelectors = hasContext && selectorExecutor(contextSelectors);
        if ('minimumSelectorsExpected' in bundleDescriptor && contextSelectors.length < bundleDescriptor.minimumSelectorsExpected) {
            throw new Error(`The bundle ${name} expects at least ${bundleDescriptor.minimumSelectorsExpected} selectors to be provided.`);
        }
        const mapStateToProps = bundleDescriptor.selector ? (firstState, firstProps) => {
            const firstContext = hasContext ? executeContextSelectors(firstProps) : [];
            let firstValue = bundleDescriptor.selector(firstState, ...firstContext);
            if (typeof firstValue !== 'function') { // no factory, fully memoized first call
                return memoizeFirstMapStateToProps(bundleDescriptor.selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext);
            }
            if (hasContext) { // factory, but memoize first context if we can. Passing in a new object as firstState means we'll never return a memoized firstValue
                return memoizeFirstMapStateToProps(firstValue, executeContextSelectors, {}, firstProps, null, firstContext);
            }
            // factory, no context, no memoization
            return state => firstValue(state);
        } : null;
        const mapDispatchToProps = typeof bundleDescriptor.actions === 'function' ?
            hasContext ?
                (dispatch, props) => bundleDescriptor.actions(dispatch, ...executeContextSelectors(props)) :
                dispatch => bundleDescriptor.actions(dispatch) :
            bundleDescriptor.actions ?
                dispatch => bindActionCreators(bundleDescriptor.actions, dispatch) :
                null;
        return {
            mapStateToProps,
            mapDispatchToProps,
        };
    };
    if (name) {
        Object.defineProperty(Bundle, 'name', {
            writable: false,
            enumerable: false,
            configurable: true,
            value: `Bundle(${name})`
        });
    }
    return Bundle;
};

export const bundle = (bundleDescriptor, name) => {
    checkBundleDescriptor(bundleDescriptor, "bundleDescriptor", "bundle(bundleDescriptor)", "");
    return internalBundle(bundleDescriptor, name);
};

export const bundleAll = bundleMap => {
    const bundleNames = checkBundleMapAndGetNames(bundleMap);
    return bundleNames.reduce((o,name) => {
        const bundleDescriptor = bundleMap[name];
        checkBundleDescriptor(bundleDescriptor, `bundleMap['${name}']`, "bundleAll(bundleMap)", "each property to be ");
        o[name] = internalBundle(bundleDescriptor, name);
        return o;
    }, {});
};