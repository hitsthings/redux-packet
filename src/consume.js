import { bindActionCreators } from "redux";
import { connect } from 'react-redux';

/*
//usage of consume
consume([
    users.fromGroup(props => props.group),
    cards.fromProject(props => props.project, props => props.filterTerm),
])(({ users = [], usersLoading, usersAllLoaded, usersLoadingErrors, cards, cardsLoading, cardsLoadingErrors, loadUsers, addUserToGroup}) => (
    <div>
        <button disabled={usersLoading || usersLoadingErrors} onClick={usersLoading || usersLoadingErrors ? null : loadUsers}>
            Load users
        </button>
        {users.map(user => (
            <p key={user.id}>{user.name}</p>
        ))}
    </div>
))
*/
const concatKeys = (arr, obj) => arr.concat(Object.keys(obj));
const nonUniqueFilter = (value, i, values) => values.lastIndexOf(value) !== i;
const defaultMapPacketsToProps = (...values) => {
    const nonUniqueProps = values.reduce(concatKeys,[]).filter(nonUniqueFilter);
    if (nonUniqueProps.length) {
        throw new Error("Props are not unique. " +
            "Pass in mapPacketsToProps to deduplicate them. " +
            "Duplicate names: " + nonUniqueProps.join(','));
    }
    return Object.assign({}, ...values);
};
const defaultMergeProps = (packetProps, ownProps) => ({
    ...ownProps,
    ...packetProps
})

const noop = (a) => {};
const usesContext = fn => typeof fn === 'function' && (fn.dependsOnOwnProps == null ? fn.length !== 1 : Boolean(fn.dependsOnOwnProps));
const stateUsesContext = packet => usesContext(packet.mapStateToProps);

export const makeMergePropsOnePacket = (mapPacketsToProps, mergeProps) => {
    if (!mapPacketsToProps) {
        if (!mergeProps) {
            return undefined;
        }
        return (stateProps, dispatchProps, ownProps) => mergeProps(
            {
                ...stateProps,
                ...dispatchProps
            },
            ownProps
        );
    }
    if (!mergeProps) {
        return (stateProps, dispatchProps, ownProps) => ({
            ...ownProps,
            ...mapPacketsToProps({
                ...stateProps,
                ...dispatchProps
            })
        });
    }
    return (stateProps, dispatchProps, ownProps) => mergeProps(
        mapPacketsToProps({
            ...stateProps,
            ...dispatchProps
        }),
        ownProps
    );
};

export const consumeOnePacket = (
    packet,
    mapPacketsToProps,
    mergeProps,
    options
) => connect(
    packet.mapStateToProps || undefined,
    packet.mapDispatchToProps || undefined,
    makeMergePropsOnePacket(mapPacketsToProps, mergeProps),
    options
);

const mapAllToPackets = mappers => {
    const anyContext = mappers.filter(Boolean).some(usesContext);
    return anyContext
        ? (stateOrDispatch, props) => ({
            packets: mappers.map(mapper => mapper(stateOrDispatch, props))
        })
        : stateOrDispatch => ({
            packets: mappers.map(mapper => mapper(stateOrDispatch))
        });
};

export const makeMapStateToProps = (packets, withState) =>
    (withState.length || undefined) && (
        withState.some(stateUsesContext)
            ? (state, props) => mapAllToPackets(packets.map(b => b.mapStateToProps ? b.mapStateToProps(state, props) : noop))
            : state => mapAllToPackets(packets.map(b => b.mapStateToProps ? b.mapStateToProps(state) : noop))
    );

export const makeMapDispatchToProps = (packets, withDispatch) =>
    (withDispatch.length || undefined) && (
        mapAllToPackets(packets.map(packet =>
            packet.mapDispatchToProps
                ? typeof packet.mapDispatchToProps === 'function'
                    ? packet.mapDispatchToProps
                    : dispatch => bindActionCreators(packet.mapDispatchToProps, dispatch)
                : noop
        ))
    );

export const makeMergeProps = (mapPacketsToProps, mergeProps) => {
    if (!mapPacketsToProps) {
        mapPacketsToProps = defaultMapPacketsToProps;
    }
    if (!mergeProps) {
        mergeProps = defaultMergeProps;
    }
    return (stateProps, dispatchProps, ownProps) => {
        const propsByPacket = stateProps.packets.map((packetState, i) => {
            const packetDispatch = dispatchProps.packets[i];
            return {
                ...packetState,
                ...packetDispatch
            };
        });
        const packetProps = mapPacketsToProps(...propsByPacket);
        return mergeProps(packetProps, ownProps);
    };
};

export const consumeMultiplePackets = (
    packets,
    mapPacketsToProps,
    mergeProps,
    options
) => connect(
    makeMapStateToProps(packets, packets.filter(b => b.mapStateToProps)),
    makeMapDispatchToProps(packets, packets.filter(b => b.mapDispatchToProps)),
    makeMergeProps(mapPacketsToProps, mergeProps),
    options
);

export const consume = (
    packets,
    mapPacketsToProps,
    mergeProps,
    options
) => {
    if (packets && (packets.mapStateToProps || packets.mapDispatchToProps)) {
        packets = [packets];
    }
    if (!Array.isArray(packets) || packets.length === 0) {
        throw new TypeError('At least one packet must be passed to consume()')
    }

    if (packets.length === 1) {
        return consumeOnePacket(packets[0], mapPacketsToProps, mergeProps, options);
    }
    return consumeMultiplePackets(packets, mapPacketsToProps, mergeProps, options);
};
