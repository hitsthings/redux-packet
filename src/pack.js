import { bindActionCreators } from "redux";

/*
//usage of pack
packAll({
    fromContext: {
        selector: (state, contextA, contextB) => `itemsInContext`, // or a function returning this to make one per component
        actions: (contextA, contextB) => `actionsForContext`,
    }
})
const users = packAll({
    fromGroup: {
        selector: () => createStructuredSelector({
            users: usersForGroupSelector,
            usersLoading: usersLoadingForGroupSelector,
            usersAllLoaded: usersAllLoadedForGroupSelector,
            usersLoadingErrors: usersLoadingErrorsForGroupSelector,
        }),
        actions: (dispatch, group) => ({
            loadUsers: () => dispatch(loadUsersForGroup(group)),
            addUserToGroup: user => dispatch(addUserToGroup(user, group)),
        }), // OR a raw object like actions: { actionA: actionCreatorA, actionB: actionCreatorB }
    },
    byPermission: { ... }
})
*/

const memoizeFirstMapStateToProps = (selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext) => {
    if (executeContextSelectors) {
        let inner = (state, props) => {
            inner = (state, props) => selector(state, ...executeContextSelectors(props));
            if (props !== firstProps) { // nothing salvageable
                return inner(state, props); 
            }
            if (state !== firstState) { 
                return selector(state, ...firstContext); // same props, can memoize contexts
            }
            return firstValue;// same props and state, can return first value
        };
        return (state, props) => inner(state, props);
    }
    let inner = state => {
        inner = selector;
        return state === firstState ?
            firstValue :
            inner(state);
    };
    return state => inner(state);
};

const checkPacketDescriptor = (packetDescriptor, name, calledFnName, argExpectation) => {
    if (typeof packetDescriptor !== 'object' || packetDescriptor === null) {
        throw new TypeError(`${calledFnName} expects ${argExpectation}an object, but ${name} was not.`);
    }
    if (packetDescriptor.selector && typeof packetDescriptor.selector !== 'function') {
        throw new TypeError(`${calledFnName} expects selector to be a function, but ${name}.selector was not.`);
    }
    if (packetDescriptor.actions && (typeof packetDescriptor.actions !== 'function' && typeof packetDescriptor.actions !== 'object')) {
        throw new TypeError(`${calledFnName} expects actions to be a function or object, but ${name}.actions was not.`);
    }
};

const checkPacketMapAndGetNames = packetMap => {
    if (typeof packetMap !== 'object' || packetMap === null) {
        throw new TypeError('packAll(packetMap) expects packetMap to be an object.');
    }
    if (Array.isArray(packetMap)) {
        throw new TypeError('packAll(packetMap) expects packetMap to be an object, not an Array.');
    }
    const packetNames = Object.keys(packetMap);
    if (packetNames.length === 0) {
        throw new TypeError('packAll(packetMap) expected to receive an object with properties, but found no properties.')
    }
    return packetNames;
};

const selectorExecutor = selectors => {
    selectors = selectors.map(selector => typeof selector === 'function' ? selector : () => selector);
    return props => selectors.map(selector => selector(props));
};

const internalPack = (packetDescriptor, name) => {
    const PacketMaker = (...contextSelectors) => {
        const hasContext = contextSelectors.length !== 0;
        const executeContextSelectors = hasContext && selectorExecutor(contextSelectors);
        if ('minimumSelectorsExpected' in packetDescriptor && contextSelectors.length < packetDescriptor.minimumSelectorsExpected) {
            throw new Error(`The packet ${name} expects at least ${packetDescriptor.minimumSelectorsExpected} selectors to be provided.`);
        }
        const mapStateToProps = packetDescriptor.selector ? (firstState, firstProps) => {
            const firstContext = hasContext ? executeContextSelectors(firstProps) : [];
            let firstValue = packetDescriptor.selector(firstState, ...firstContext);
            if (typeof firstValue !== 'function') { // no factory, fully memoized first call
                return memoizeFirstMapStateToProps(packetDescriptor.selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext);
            }
            if (hasContext) { // factory, but memoize first context if we can. Passing in a new object as firstState means we'll never return a memoized firstValue
                return memoizeFirstMapStateToProps(firstValue, executeContextSelectors, {}, firstProps, null, firstContext);
            }
            // factory, no context, no memoization
            return state => firstValue(state);
        } : null;
        const mapDispatchToProps = typeof packetDescriptor.actions === 'function' ?
            hasContext ?
                (dispatch, props) => packetDescriptor.actions(dispatch, ...executeContextSelectors(props)) :
                dispatch => packetDescriptor.actions(dispatch) :
                packetDescriptor.actions ?
                dispatch => bindActionCreators(packetDescriptor.actions, dispatch) :
                null;
        return {
            mapStateToProps,
            mapDispatchToProps,
        };
    };
    if (name) {
        Object.defineProperty(PacketMaker, 'name', {
            writable: false,
            enumerable: false,
            configurable: true,
            value: `PacketMaker(${name})`
        });
    }
    return PacketMaker;
};

export const pack = (packetDescriptor, name) => {
    checkPacketDescriptor(packetDescriptor, "packetDescriptor", "pack(packetDescriptor)", "");
    return internalPack(packetDescriptor, name);
};

export const packAll = packetMap => {
    const packetNames = checkPacketMapAndGetNames(packetMap);
    return packetNames.reduce((o,name) => {
        const packetDescriptor = packetMap[name];
        checkPacketDescriptor(packetDescriptor, `packetMap['${name}']`, "packAll(packetMap)", "each property to be ");
        o[name] = internalPack(packetDescriptor, name);
        return o;
    }, {});
};