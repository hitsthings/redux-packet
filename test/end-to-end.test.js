import React, { Fragment } from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';
import { createStore } from 'redux';
import { connect, Provider } from 'react-redux';

import { packAll } from '../src/pack';
import { consume } from '../src/consume';

function immutableSet(o, path, v) {
    path = path.split('.');
    const modified = path.reduce((out, p) => out.concat(out[out.length - 1][p] || {}), [o]);
    const newObjects = modified.map((o, i) => i === modified.length - 1 ? v : {...o});
    path.forEach((p, i) => {
        newObjects[i][p] = newObjects[i+1];
    })
    return newObjects[0];
}
const store = createStore(function(state = {}, action) {
    if (action.type === '___reset') return {};
    if (action.type.startsWith('@')) return state;
    return immutableSet(state, action.type, action.payload);
});
const set = (path, value) => ({
    type: path,
    payload: value
});
const get = (state, path) => path.split('.').reduce((o,p) => o && o[p], state);

const cache = new Map();
const cacheGetOr = (key, supply) => {
    if (!cache.has(key)) {
        cache.set(key, supply());
    }
    return cache.get(key);
};

const inContext = (state, path) => {
    const context = get(state, path)
    // cache against the object or undefined
    return cacheGetOr(context, () => ({
        users: get(context, `users`),
        usersLoading: get(context, `usersLoading`),
        usersLoadingErrors: get(context, `usersLoadingErrors`),
        usersLoaded: get(context, `usersLoaded`)
    }));
};
const loadForContext = (dispatch, context) => {
    dispatch(set(`${context}.usersLoading`, true));
    setTimeout(() => {
        dispatch(set(context, {
            users: ['Alice', 'Bob', 'Eve'],
            usersLoading: false,
            usersLoadingErrors: [],
            usersLoaded: true,
        }));                    
    }, 100);

};

const users = packAll({
    forGroup: {
        selector: (state, group) => {
            return inContext(state, `groups.${group}`);
        },
        actions: (dispatch, group) => ({
            loadUsersForGroup: () => loadForContext(dispatch, `groups.${group}`)
        })
    },
    forProject: {
        selector: (state, project) => {
            return inContext(state, `projects.${project}`);
        },
        actions: (dispatch, project) => ({
            loadUsersForProject: () => loadForContext(dispatch, `projects.${project}`)
        })
    },
    forProjectInGroup: {
        selector: (state, group, project) => {
            return inContext(state, `groups.${group}.projects.${project}`);
        },
        actions: (dispatch, group, project) => ({
            loadUsers: () => loadForContext(dispatch, `groups.${group}.projects.${project}`)
        })
    }
});

const UsersList = ({ users = [], usersLoading, usersLoadingErrors = [], usersLoaded, loadUsers }) => (
    <Fragment>
        {usersLoading && <span className="loading">Loading...</span>}
        {usersLoadingErrors.length &&
            <ul className="errors">
                {usersLoadingErrors.map(error => (
                    <li key={error}>{error}</li>
                ))}
            </ul>
        }
        {users.length &&
            <ul className="users">
                {users.map(u => (
                    <li key={u}>{u}</li>
                ))}
            </ul>
        }
        {usersLoading || usersLoaded || <button onClick={loadUsers}>Load users</button>}
    </Fragment>
);

const GroupUsersList = consume([
    users.forGroup(props => props.groupId)
], usersForGroup => ({
    loadUsers: usersForGroup.loadUsersForGroup,
    ...usersForGroup
}))(UsersList);

const ProjectUsersList = consume([
    users.forProject(props => props.projectId)
], usersForProject => ({
    loadUsers: usersForProject.loadUsersForProject,
    ...usersForProject
}))(UsersList);

const GroupProjectUsersList = consume([
    users.forProjectInGroup(props => props.groupId, props => props.projectId)
])(UsersList);

const GroupProjectUsersList2 = consume([
    users.forProject(props => props.projectId),
    users.forProjectInGroup(props => props.groupId, props => props.projectId)
], (a, b) => b, null)(UsersList);

const DuplicatePropNames = consume([
    users.forProject(props => props.projectId),
    users.forProjectInGroup(props => props.groupId, props => props.projectId)
])(UsersList);

const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

const withSupressedLogging = fn => (...args) => {
    const { log, warn, error } = console;
    console.log = console.warn = console.error = () => {};
    try {
        return fn(...args);
    } finally {
        console.log = log;
        console.warn = warn;
        console.error = error;
    }
}

describe('consume([packAll({...})])', () => {
    beforeEach(() => {
        store.dispatch({ type: '___reset' })
    })
    const testElement = element => async () => {
        const container = document.createElement('div');
        ReactDOM.render(<Provider store={store}>{element}</Provider>, container);
        expect(container.children.length).toBe(1);
        expect(container.querySelector('button')).not.toBe(null);

        ReactTestUtils.Simulate.click(container.querySelector('button'));
        expect(container.children.length).toBe(1);
        expect(container.querySelector('span')).not.toBe(null);
        
        await wait(100);
        expect(container.children.length).toBe(1);
        expect(container.querySelector('.users')).not.toBe(null);
    };
    it('can handle users in a group', testElement(<GroupUsersList groupId={2} />));
    it('can handle users in a project', testElement(<ProjectUsersList projectId={2} />));
    it('can handle users in a project in a group', testElement(<GroupProjectUsersList groupId={3} projectId={2} />));
    it('can handle multiple packets', testElement(<GroupProjectUsersList2 groupId={3} projectId={2} />));
    it('throws on duplicate props', () => {
        expect(withSupressedLogging(() => {
            ReactDOM.render(<Provider store={store}><DuplicatePropNames /></Provider>, document.createElement('div'));
        })).toThrowErrorMatchingSnapshot();
    });

    it("doesn't rerender if relevant state hasn't changed", () => {
        const mapPacketsToProps = jest.fn((forGroup, forProject) => forGroup);

        const MyComponent = jest.fn().mockReturnValue(<div>Rendered</div>);
        const StatefulMyComponent = consume(
            [users.forGroup(() => 2), users.forProject(() => 2)],
            mapPacketsToProps
        )(MyComponent);

        const container = document.createElement('div');
        ReactDOM.render(<Provider store={store}><StatefulMyComponent /></Provider>, container);
        expect(mapPacketsToProps).toBeCalledTimes(1);
        expect(MyComponent).toBeCalledTimes(1);

        store.dispatch(set("unrelated.path", 42));
        expect(mapPacketsToProps).toBeCalledTimes(1);
        expect(MyComponent).toBeCalledTimes(1);

        store.dispatch(set(`groups.2.usersLoading`, true));
        expect(mapPacketsToProps).toBeCalledTimes(2);
        expect(MyComponent).toBeCalledTimes(2);
    });
});