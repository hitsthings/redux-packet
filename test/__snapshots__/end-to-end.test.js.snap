// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consume([packAll({...})]) throws on duplicate props 1`] = `"Props are not unique. Pass in mapPacketsToProps to deduplicate them. Duplicate names: users,usersLoading,usersLoadingErrors,usersLoaded"`;
